src/types/list.rs

```diff
@@ -462,29 +483,119 @@ impl<'py> PyListMethods<'py> for Bound<'py, PyList> {
     }
 }
 
+// New types for type checking when using BoundListIterator associated methods, like
+// BoundListIterator::next_unchecked.
+struct Index(usize);
+struct Length(usize);
+
 /// Used by `PyList::iter()`.
 pub struct BoundListIterator<'py> {
     list: Bound<'py, PyList>,
-    index: usize,
-    length: usize,
+    index: Index,
+    length: Length,
+}
+
+/// Helper to manage mutable borrows below
+macro_rules! split_borrow {
+    ($instance:expr, $index:ident, $length:ident, $list:ident) => {
+        let Self {
+            ref mut $index,
+            ref mut $length,
+            ref $list,
+        } = $instance;
+    };
 }
 
 impl<'py> BoundListIterator<'py> {
     fn new(list: Bound<'py, PyList>) -> Self {
-        let length: usize = list.len();
-        BoundListIterator {
+        Self {
+            index: Index(0),
+            length: Length(list.len()),
             list,
-            index: 0,
-            length,
         }
     }
 
-    unsafe fn get_item(&self, index: usize) -> Bound<'py, PyAny> {
-        #[cfg(any(Py_LIMITED_API, PyPy, Py_GIL_DISABLED))]
-        let item = self.list.get_item(index).expect("list.get failed");
-        #[cfg(not(any(Py_LIMITED_API, PyPy, Py_GIL_DISABLED)))]
-        let item = self.list.get_item_unchecked(index);
-        item
+    /// # Safety
+    ///
+    /// On the free-threaded build, caller must verify they have exclusive
+    /// access to the list by holding a lock or by holding the innermost
+    /// critical section on the list.
+    #[inline]
+    #[cfg(all(not(Py_LIMITED_API), not(PyPy)))]
+    unsafe fn next_unchecked(
+        index: &mut Index,
+        length: &mut Length,
+        list: &Bound<'py, PyList>,
+    ) -> Option<Bound<'py, PyAny>> {
+        let length = length.0.min(list.len());
```

Comment:
@mejrs if a closure updates the length and the old length stores in the iterator is out-of-bounds, this check means that `next_unchecked()` returns `None` and the iteration terminates.

 We hold a critical section so other threads can't modify the list between here and the next `get_item_unchecked` call.

Does that make sense? Are you worried about other scenarios?

In any case, I'll try to add a test where the closure modifies the list to see what happens....

Reply:
See `test_iter_fold_out_of_bounds` added in the last commit.